### PR TITLE
added filter options to TX history

### PR DESCRIPTION
**Pull request by allienworks**
- added TX filtering options (by type and by narration)
- clean-up of old History styles

FYI, filtering by narration can be hidden now. I felt like it could be a nice addition to filtering, but it's not necessarily needed at the moment (I don't know how much work would be needed for its implementation).

### TODO

- [x] filter by TX type (public / blind / anon)
- [x] show number of results

### CODE REVIEW

**Comment by pciavald** [comment]
amazing, thank you very much !

**Comment by pciavald** [comment]
for more filters : particl/particl-core#12 in the help section

**Comment by kewde** [comment]
The pagination requires the total amount of transaction that apply for that specificaties filter. We're not loading all txs at the moment so we have no estimation for the pagination?

**Comment by pciavald** [comment]
we could fetch without limit to setup all pages and iterate in the result

**Comment by pciavald** [comment]
@allienworks can you please change the `Filter narration` selector from a checkbox to a radio button ? it doesn't make sense to be able to select both with and without narration :)

**Comment by allienworks** [comment]
@pciavald Hehe, you're right about that. Switched to radios already. I'd unselect the radios by default (they are selected now, no idea why).

File: src/app/wallet/wallet/history/history.component.html [hunk 1]
```diff
@@ -1,24 +1,71 @@
...
+    <p class="widget-help">
+      Transaction filter gives you the most granular control over displayed transactions.
+    </p>
+
+    <div class="section filter-by-narration">
+      <div class="subtitle">Filter by narrations</div>
+      <mat-card class="filter-card">
+        <mat-radio-group class="radio-group">
+          <!-- TODO: unselect them by default (no idea where to define it, tbh): -->
+          <mat-radio-button class="filter-option" *ngFor="let filter of narration_filter" [value]="season">
```
Comment (pciavald) [inline src/app/wallet/wallet/history/history.component.html]: @allienworks what is `season` ?

**Comment by pciavald** [comment]
@kewde do you understand why this fails ? `ng test` and `ng lint` pass locally

File: src/app/wallet/wallet/history/history.component.html [hunk 2]
```diff
@@ -1,24 +1,50 @@
...
+<app-header class="disable-select"></app-header>
+
+<div fxLayout="row" fxLayoutGap="30px" class="container-block">
+
+  <div class="filter" fxFlex="0 0 200px">
+
+    <div class="section filter-by-type">
+      <div class="subtitle">Filter by type</div>
+      <mat-card class="filter-card">
+        <mat-selection-list #types dense class="filter-list no-padding">
```
Comment (pciavald) [inline src/app/wallet/wallet/history/history.component.html]: @allienworks this seems to trigger an error in the console, can you look into it ?

Comment (pciavald) [reply src/app/wallet/wallet/history/history.component.html]: https://github.com/angular/material2/blob/master/src/demo-app/list/list-demo.html

**Comment by pciavald** [comment]
Missing fields in the GUI : 
- skip : int > 0
- watch-only transactions : checkbox
- search : text
- type : checkbox list [all, standard, anonymous, blind]
- sort : dropdown / radio list [time, address, category, amount, confirmations, txid]

**Comment by pciavald** [comment]
### can't interact with the tabs
I'm trying to catch a click event on the tabs but can't seem to be able to do it. Maybe embed a button ?

### error when selecting an element from the sort menu
```
ERROR TypeError: Cannot read property 'nativeElement' of undefined
    at MatSelect.webpackJsonp.../../../material/esm5/select.es5.js.MatSelect._onFadeInDone (file:///home/pciavald/Documents/partgui/dist/vendor.bundle.js:136278:19)
    at Object.eval [as handleEvent] (ng:///MatSelectModule/MatSelect.ngfactory.js:70:27)
    at handleEvent (file:///home/pciavald/Documents/partgui/dist/vendor.bundle.js:98548:172)
    at callWithDebugContext (file:///home/pciavald/Documents/partgui/dist/vendor.bundle.js:100033:42)
    at Object.debugHandleEvent [as handleEvent] (file:///home/pciavald/Documents/partgui/dist/vendor.bundle.js:99620:12)
    at dispatchEvent (file:///home/pciavald/Documents/partgui/dist/vendor.bundle.js:94997:25)
    at file:///home/pciavald/Documents/partgui/dist/vendor.bundle.js:95611:38
    at file:///home/pciavald/Documents/partgui/dist/vendor.bundle.js:144923:25
    at Array.forEach (native)
    at file:///home/pciavald/Documents/partgui/dist/vendor.bundle.js:144921:53
```

### error when clicking the input
```
ERROR TypeError: _co.selectInput is not a function
    at Object.eval [as handleEvent] (HistoryComponent.html:24)
    at handleEvent (vendor.bundle.js:98548)
    at callWithDebugContext (vendor.bundle.js:100033)
    at Object.debugHandleEvent [as handleEvent] (vendor.bundle.js:99620)
    at dispatchEvent (vendor.bundle.js:94997)
    at vendor.bundle.js:95611
    at HTMLElement.<anonymous> (vendor.bundle.js:148181)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (polyfills.bundle.js:2964)
    at Object.onInvokeTask (vendor.bundle.js:89913)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (polyfills.bundle.js:2963)
```

also, that field searches for narrations that are not displayed #496

**Comment by pciavald** [comment]
@allienworks do you have the time to handle these or should we ask to someone else ? I'm trying to keep away from pure front as I'm slower on these tasks.

**Comment by allienworks** [comment]
@pciavald yeah sorry, I don't think I'll be able to help with this. I've looked over it already, no idea what might be wrong there. Moreover I just copy-pasted the example code from the docs (so no reason for it not to work IMO).

**Comment by pciavald** [comment]
@allienworks lost cross when pulled dev
screenshot_20171215_191747

**Comment by pciavald** [comment]
@anandsinghparihar 
- can you check why the paginator's total items number is not updated when updating ?
- can you set an input so I can change the selected tab from the TS file ?
Thank you :)

**Comment by anandsinghparihar** [comment]
I've created a variable called `selectedTab` for changing tabs from component you can use it. and now paginator's total items number is updated when updating.

**Comment by allienworks** [comment]
@pciavald I fixed the missing icon.

Btw dunno if you know about it, but when switching between tabs, it seem to remember page number you're on. So if I have 10 pages of stakes and 2 pages of balance transfers; and I go to 5th page of stakes, then switch to balance transfer tab, I get "no transactions yet" error.

**Comment by pciavald** [comment]
@anandsinghparihar the variable for selecting tabs does not seem to work, even pre-setting it to 1 does not select the right tab. The paginator still show the total amount of transactions, not the latest update.

File: src/app/wallet/wallet/history/history.component.ts [hunk 3]
```diff
@@ -1,28 +1,88 @@
...
+
+  filters: any = {
+    category: undefined,
+    search:   undefined,
+    sort:     undefined,
+    type:     undefined
+  };
+
+  // default tab is zero
+  public selectedTab: number = 0;
```
Comment (pciavald) [inline src/app/wallet/wallet/history/history.component.ts]: this does not seem to change the tab. try with other numbers to test

File: src/app/wallet/wallet/history/history.component.ts [hunk 4]
```diff
@@ -1,28 +1,88 @@
...
+  constructor() {
+    this.default();
+  }
+
+  ngOnInit() {
+    /* may be used if we concatenate some filters http://bit.ly/2Buav9B */
+  }
+
+  default() {
+    this.selectedTab = 0;
```
Comment (pciavald) [inline src/app/wallet/wallet/history/history.component.ts]: used here to clear filters

File: src/app/wallet/wallet/shared/transaction.service.ts [hunk 5]
```diff
@@ -83,45 +94,51 @@ export class TransactionService {
...
+      this.log.d(`rpc_loadTransactions_success, supposedly tx per page: ${this.MAX_TXS_PER_PAGE}`);
+      this.log.d(`rpc_loadTransactions_success, real tx per page: ${txResponse.length}`);
+
+      if (txResponse.length !== this.MAX_TXS_PER_PAGE) {
+        this.log.er(`rpc_loadTransactions_success, TRANSACTION COUNTS DO NOT MATCH (maybe last page?)`);
+      }
+
+      this.deleteTransactions();
+      txResponse.forEach((tx) => {
+        this.addTransaction(tx);
```
Comment (pciavald) [inline src/app/wallet/wallet/shared/transaction.service.ts]: maybe the paginator should be updated here ?

**Comment by pciavald** [comment]
@anandsinghparihar rebased dev :)

File: src/app/wallet/wallet/history/history.component.ts (hunk 4 above)
Comment (anandsinghparihar) [reply src/app/wallet/wallet/history/history.component.ts]: @pciavald  Do you really think we should change tab while clearing filter?  (filter is  not a tab !!)

File: src/app/wallet/wallet/history/history.component.ts (hunk 3 above)
Comment (anandsinghparihar) [reply src/app/wallet/wallet/history/history.component.ts]: you can change no zero to 1 tab would change.

File: src/app/wallet/wallet/history/history.component.ts (hunk 4 above)
Comment (anandsinghparihar) [reply src/app/wallet/wallet/history/history.component.ts]: I just reset tabs on filter clear.

File: src/app/wallet/wallet/shared/transaction.service.ts (hunk 5 above)
Comment (anandsinghparihar) [reply src/app/wallet/wallet/shared/transaction.service.ts]: tx count is not updated after API call so could you please confirm me how can I get updated tx count after filter applied?

Comment (anandsinghparihar) [reply src/app/wallet/wallet/shared/transaction.service.ts]: Please update tx record state is not updated after getting filter record. please update state then pagination will be work fine.

**Comment by pciavald** [comment]
clear tabs / paginator
peek 21-12-2017 11-20

**Comment by anandsinghparihar** [comment]
Hi @pciavald , last day my commit was not pushed due to
`
! [remote rejected] enhancement/tx-filtering -> enhancement/tx-filtering (push declined due to email privacy restrictions) error: failed to push some refs to 'git@github.com:particl/partgui.git`

tab refresh after the clear filter. only we need tx total count update while we filter record (tx total count is not update after filter data)

File: src/app/wallet/wallet/shared/transaction.service.ts (hunk 5 above)
Comment (pciavald) [reply src/app/wallet/wallet/shared/transaction.service.ts]: @anandsinghparihar `txResponse.length` is the new total

Comment (anandsinghparihar) [reply src/app/wallet/wallet/shared/transaction.service.ts]: I guess `txResponse.length`  is returns total item per page?

**Comment by anandsinghparihar** [comment]
@pciavald could you please add a pram into API to get a count on filter !!

**Comment by pciavald** [comment]
https://github.com/particl/particl-core/issues/20

**Comment by pciavald** [comment]
waiting for https://github.com/particl/particl-core/pull/21

File: src/app/wallet/wallet/shared/transaction.service.ts [hunk 6]
```diff
@@ -80,48 +94,63 @@ export class TransactionService {
...
-            this.addTransaction(tx);
-          });
-
-        this.loading = false;
-        this.log.d(`rpc_update, txs array: ${this.txs.length}`);
+    .subscribe((txResponse: Array<Object>) => {
+
+      if (justCount) {
+        this.log.d(`number of transactions after filter: ${txResponse.length}`);
+        // TODO: update paginator length with txResponse.length
```
Comment (pciavald) [inline src/app/wallet/wallet/shared/transaction.service.ts]: @anandsinghparihar you can update paginator here :)

**Comment by anandsinghparihar** [comment]
Hi @pciavald  this PR  is ready fore review.

**Pull request merged**